src/test/test-dirent-util.c

```diff
@@ -0,0 +1,211 @@
+        assert_se(dirent_is_file(de_reg) == true);
+        assert_se(dirent_is_file(de_lnk) == true);
+        assert_se(dirent_is_file(&de_unknown) == true);
+
+        /* Test for hidden files */
+        assert_se(dirent_is_file(de_dot) == false);
+        assert_se(dirent_is_file(de_bak) == false);
+        assert_se(dirent_is_file(de_tilda) == false);
+
+        closedir(dir);
+}
+
+TEST (test_dirent_is_file_with_suffix) {
+        _cleanup_(rm_rf_physical_and_freep) char *t = NULL;
+        const char *name, *dotfile, *name_alias, *dotdot, *chr;
+        const struct dirent *de_reg, *de_lnk, *de_dot, *de_dotdot, *de_chr;
+        DIR *dir;
+
+        static const struct dirent de_unknown = {
+                .d_type = DT_UNKNOWN,
+                .d_name = "test_unknown",
+        };
+
+        assert_se(mkdtemp_malloc(NULL, &t) >= 0);
+
+        name = strjoina(t, "/test.txt");
+        dotfile = strjoina(t, "/.hidden_file");
+        dotdot = strjoina(t, "/..dotdot");
+        chr = strjoina(t, "/test_chr");
+        name_alias = strjoina(t, "/test_link");
+
+        assert_se(touch(name) >= 0);
+        assert_se(touch(dotfile) >= 0);
+        assert_se(touch(dotdot) >= 0);
+        assert_se(mknod(chr, 0775 | S_IFCHR, makedev(0, 0)) >= 0);
+
+        if (symlink(name, name_alias) < 0) {
+                assert_se(IN_SET(errno, EINVAL, ENOSYS, ENOTTY, EPERM));
+                log_tests_skipped_errno(errno, "symlink() not possible");
+        }
+
+        dir = opendir(t);
+        if (dir == NULL) {
+                log_error_errno(errno, "Failed to open directory '%s': %m", t);
+                exit(EXIT_FAILURE);
+        }
+
+        rewinddir(dir);
+        while ((de_reg = readdir_ensure_type(dir)) != NULL) {
+                if (strcmp(de_reg->d_name, "test.txt") == 0)
+                        break;
+        }
+
+        rewinddir(dir);
+        while ((de_lnk = readdir_ensure_type(dir)) != NULL) {
+                if (strcmp(de_lnk->d_name, "test_link") == 0)
+                        break;
+        }
+
+        rewinddir(dir);
+        while ((de_dot = readdir_ensure_type(dir)) != NULL) {
+                if (strcmp(de_dot->d_name, ".hidden_file") == 0)
+                        break;
+        }
+
+        rewinddir(dir);
+        while ((de_dotdot = readdir(dir)) != NULL) {
+                if (strcmp(de_dotdot->d_name, "..dotdot") == 0)
+                        break;
+        }
+
+        rewinddir(dir);
+        while ((de_chr = readdir(dir)) != NULL) {
+                if (strcmp(de_chr->d_name, "test_chr") == 0)
+                        break;
+        }
+
+        /* Test when d_type is not DT_REG, DT_LNK, or DT_UNKNOWN */
+        assert_se(!dirent_is_file_with_suffix(de_chr, NULL));
+
+        /* Test when suffix is NULL */
+        assert_se(dirent_is_file_with_suffix(de_reg, NULL) == true);
+        assert_se(dirent_is_file_with_suffix(de_lnk, NULL) == true);
```

Comment:
as per the check above, the symlilnk might fail but the test would continue, and will then abort as de_lnk is null. Check that it isn't null before performing this check